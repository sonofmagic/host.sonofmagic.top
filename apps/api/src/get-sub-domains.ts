interface Cert {
  issuer_ca_id: number
  issuer_name: string
  common_name: string
  name_value: string
  id: number
  entry_timestamp: string
  not_before: string
  not_after: string
  serial_number: string
  result_count: number
}

export async function getSubDomains(domain: string): Promise<string[]> {
  const url = `https://crt.sh/?q=%25.${domain}&output=json`
  try {
    const response = await fetch(url)
    if (!response.ok) {
      throw new Error(`Failed to fetch subdomains for ${domain}: ${response.statusText}`)
    }
    const data = await response.json() as any
    const subDomains = data.reduce((acc: string[], cert: Cert) => {
      const subDomain = cert.name_value.split('\n')
      subDomain.forEach((domain) => {
        const originDomain = domain.replace(`*.`, '')
        if (!acc.includes(originDomain)) {
          acc.push(originDomain)
        }
      })
      return acc
    }, [])
    return subDomains
  }
  catch (error) {
    console.error(error)
    return []
  }
}
