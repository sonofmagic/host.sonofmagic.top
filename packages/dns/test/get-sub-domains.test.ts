import { isCI } from 'ci-info'
import { getSubDomains } from '@/get-sub-domains'

describe.skipIf(isCI)('get-sub-domains', () => {
  it('sonofmagic.top', async () => {
    const subDomains = await getSubDomains('sonofmagic.top')
    expect(subDomains.sort()).toMatchSnapshot()
  })
})
