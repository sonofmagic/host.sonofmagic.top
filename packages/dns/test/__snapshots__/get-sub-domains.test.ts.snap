// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`get-sub-domains > sonofmagic.top 1`] = `
[
  "api.host.sonofmagic.top",
  "dev.host.sonofmagic.top",
  "docs.host.sonofmagic.top",
  "host.sonofmagic.top",
  "sonofmagic.top",
]
`;
